test/test_jit.py

```diff
@@ -2021,6 +2021,132 @@ def forward(self, input):
         with self.assertRaisesRegex(RuntimeError, "cannot re-assign"):
             m.sub = nn.Linear(5, 5)
 
+    def test_script_module_const(self):
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__(False)
+                self.b = torch.jit.Const(False)
+                self.i = torch.jit.Const(1)
+                self.c = torch.jit.Const(3.5)
+
+            @torch.jit.script_method
+            def forward(self):
+                return self.b, self.i, self.c
+
+        m = M()
+        o0, o1, o2 = m()
+        self.assertEqual(o0, 0)
+        self.assertEqual(o1, 1)
+        self.assertEqual(o2, 3.5)
+
+    def test_script_module_fail_const(self):
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__(False)
+                self.b = False
+
+            @torch.jit.script_method
+            def forward(self):
+                return self.b
+        with self.assertRaisesRegex(RuntimeError, "is not usable in a script method"):
+            M()
+
+    def test_script_module_valid_consts(self):
+        torch.jit.Const(1)
+        torch.jit.Const(1.2)
+        torch.jit.Const(False)
+        torch.jit.Const(nn.Linear(3, 4))
+        torch.jit.Const(lambda x: x)
+        self.assertTrue(type(torch.jit.Const([3, 4, 5])._value) is tuple)
+        torch.jit.Const([3, (3, 4), 5])
+        with self.assertRaisesRegex(TypeError, "is not a valid constant"):
+            torch.jit.Const(type(1))
+        with self.assertRaisesRegex(TypeError, "is not a valid constant"):
+            torch.jit.Const((3, 4, {}))
+
+    def test_script_module_for(self):
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__(False)
+                self.b = torch.jit.Const([1, 2, 3, 4])
+
+            @torch.jit.script_method
+            def forward(self):
+                sum = 0
+                for i in self.b:
+                    sum += i
+                return sum
+
+        m = M()
+        self.assertEqual(m(), 10)
+
+    def test_script_module_for2(self):
+        class Sub(torch.jit.ScriptModule):
+            def __init__(self):
+                super(Sub, self).__init__(False)
+                self.weight = nn.Parameter(torch.randn(2))
+
+            @torch.jit.script_method
+            def forward(self, thing):
+                return self.weight + thing
+
+        class M(torch.jit.ScriptModule):
+            def __init__(self):
+                super(M, self).__init__(False)
+                mods = []
+                for i in range(10):
+                    m = Sub()
+                    self.add_module('m' + str(i), m)
+                    mods.append(m)
```

Comment:
Keeping modules in regular Python lists is a very bad idea. It will break e.g. with `DataParallel` and in multiple other occasions. That's why we have `nn.ModuleList`

Reply:
In this case, they are both added as a submodule via add_module, and added to the list. Is that still a problem?

Reply:
Yes. The list can get out of sync with `_modules`.

Reply:
We can have torch.jit.const([<module_list>]) return a 'ConstModuleList' which would be a subclass of ScriptModule that also behaves like a list similar to nn.ModuleList. This would then eliminate all the "is this a submodule" checking, which I like.

Reply:
Sounds good!